vllm/model_executor/layers/sampler.py

```diff
@@ -384,6 +395,116 @@ def _apply_min_tokens_penalty(
+    logits -= frequency_penalties.unsqueeze_(dim=1) * output_bin_counts
+    logits -= presence_penalties.unsqueeze_(dim=1) * output_mask
+    return logits
+
+
+def _apply_dry(
+    logits: torch.Tensor,
+    input_token_ids: torch.Tensor,
+    output_token_ids: torch.Tensor,
+    multipliers: torch.Tensor,
+    bases: torch.Tensor,
+    allowed_lengths: torch.Tensor,
+    sequence_breakers_ids: torch.Tensor,
+    ranges: torch.Tensor,
+) -> torch.Tensor:
+    """
+    Apply Don't Repeat Yourself (DRY) sampling to the logits.
+
+    Reference: https://github.com/PygmalionAI/aphrodite-engine/blob/a3c03db7355b33c0dfd670b084e827d0aa7442d1/aphrodite/modeling/layers/sampler.py#L621-L702
+    """
+
+    VOCAB_SIZE = logits.size(-1)
+    MAX_NGRAM = 100
+
+    # Process each sequence that has a nonzero multiplier
+    applies_to = multipliers.nonzero(as_tuple=True)[0]
+    for irow in applies_to.tolist():
+        # DRY applies to input AND output tokens, so concat w/o padding tokens
+        prompt_len = (len(input_token_ids[irow]) -
+                      (input_token_ids[irow] == VOCAB_SIZE).sum().item())
+        ouput_len = (len(output_token_ids[irow]) -
+                     (output_token_ids[irow] == VOCAB_SIZE).sum().item())
+        token_seq = torch.cat((input_token_ids[irow][:prompt_len],
+                               output_token_ids[irow][:ouput_len]),
+                              dim=0)
+
+        range_limit = ranges[irow].item()
+        if range_limit:
+            token_seq = token_seq[-range_limit:]
+
+        last_token = token_seq[-1].item()
+        if last_token in sequence_breakers_ids[irow]:
+            continue  # early out for everything up to the min_ngram check?
+
+        # Build a mask of all the breaking tokens in the context
+        break_mask = torch.zeros(len(token_seq),
+                                 dtype=torch.bool,
+                                 device=logits.device)
+        for break_tok in sequence_breakers_ids[irow]:
+            break_mask.logical_or_(token_seq == break_tok)
+
+        # Find the most recent breaking token (sets ngram limit)
+        max_ngram = 0
+        for max_ngram in range(min(len(break_mask), MAX_NGRAM + 1)):
+            if break_mask[-max_ngram - 1]:
+                break
+
+        min_ngram = allowed_lengths[irow].item()
+        if max_ngram <= min_ngram:  # Too close to a break to match anything
+            continue
+
+        # If [token] is picked, what's the longest ngram that would match?
+        ngram_lens = torch.zeros(VOCAB_SIZE,
+                                 dtype=torch.int32,
+                                 device=logits.device)
+
+        # Find all instances of the last token- potential ngrams!
+        endpoint_indexes = torch.nonzero(token_seq == last_token,
+                                         as_tuple=True)[0].tolist()
+        # NOTE(alpin): This seems like the slow part.
```

Comment:
Here's an idea: Instead of matching a single token above with `nonzero`, try to match the last `allowed_length` tokens. Shorter matches don't need to be penalized by definition, and this should drastically cut down on the number of matches that must be looped over. I'm not sure how to perform subsequence matching in PyTorch, but there should be a way to do this as a vectorized operation.